pkg/dockerregistry/server/signaturedispatcher.go

```diff
@@ -0,0 +1,166 @@
+	reference imageapi.DockerImageReference
+}
+
+// SignatureDispatcher handles the GET and PUT requests for signature endpoint.
+func SignatureDispatcher(ctx *handlers.Context, r *http.Request) http.Handler {
+	signatureHandler := &signatureHandler{ctx: ctx}
+	signatureHandler.reference, _ = imageapi.ParseDockerImageReference(ctxu.GetStringValue(ctx, "vars.name") + "@" + ctxu.GetStringValue(ctx, "vars.digest"))
+	return gorillahandlers.MethodHandler{
+		"GET": http.HandlerFunc(signatureHandler.Get),
+		"PUT": http.HandlerFunc(signatureHandler.Put),
+	}
+}
+
+func (s *signatureHandler) Put(w http.ResponseWriter, r *http.Request) {
+	context.GetLogger(s.ctx).Debugf("(*signatureHandler).Put")
+	client, ok := UserClientFrom(s.ctx)
+	if !ok {
+		s.handleError(s.ctx, errcode.ErrorCodeUnknown.WithDetail("unable to get origin client"), w)
+		return
+	}
+
+	sig := signature{}
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		s.handleError(s.ctx, ErrorCodeSignatureInvalid.WithDetail(err.Error()), w)
+		return
+	}
+	if err := json.Unmarshal(body, &sig); err != nil {
+		s.handleError(s.ctx, ErrorCodeSignatureInvalid.WithDetail(err.Error()), w)
+		return
+	}
+
+	if len(sig.Type) == 0 {
+		sig.Type = imageapi.ImageSignatureTypeAtomicImageV1
+	}
+	newSig := &imageapi.ImageSignature{Content: sig.Content, Type: sig.Type}
+	newSig.Name = sig.Name
+
+	if _, err := client.ImageSignatures().Create(newSig); err != nil {
+		switch {
+		case kapierrors.IsUnauthorized(err):
+			s.handleError(s.ctx, errcode.ErrorCodeUnauthorized.WithDetail(err.Error()), w)
+		case kapierrors.IsBadRequest(err):
+			s.handleError(s.ctx, ErrorCodeSignatureInvalid.WithDetail(err.Error()), w)
+		case kapierrors.IsNotFound(err):
+			w.WriteHeader(http.StatusNotFound)
+		case kapierrors.IsAlreadyExists(err):
+			s.handleError(s.ctx, ErrorCodeSignatureAlreadyExists.WithDetail(err.Error()), w)
+		default:
+			s.handleError(s.ctx, errcode.ErrorCodeUnknown.WithDetail(fmt.Sprintf("unable to create image %s signature: %v", s.reference.String(), err)), w)
+		}
+		return
+	}
+
+	// Return just 201 with no body.
+	// TODO: The docker registry actually returns the Location header
+	w.WriteHeader(http.StatusCreated)
+	context.GetLogger(s.ctx).Debugf("(*signatureHandler).Put signature successfully added to %s", s.reference.String())
+}
+
+func (s *signatureHandler) Get(w http.ResponseWriter, req *http.Request) {
+	context.GetLogger(s.ctx).Debugf("(*signatureHandler).Get")
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+
+	client, ok := UserClientFrom(s.ctx)
+	if !ok {
+		s.handleError(s.ctx, errcode.ErrorCodeUnknown.WithDetail("unable to get origin client"), w)
+		return
+	}
+
+	if len(s.reference.ID) == 0 {
+		s.handleError(s.ctx, v2.ErrorCodeNameInvalid.WithDetail("the image ID must be specified (sha256:<digest>"), w)
+		return
+	}
+
+	image, err := client.ImageStreamImages(s.reference.Namespace).Get(s.reference.Name, s.reference.ID)
+	if err != nil {
```

Comment:
style: if you have the switch, make err == nil the first case with a no-op.  Then you don't need the if block.